include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge_sort.h

```diff
@@ -228,73 +228,414 @@ struct __merge_sort_leaf_submitter<__internal::__optional_kernel_name<_LeafSortN
+                     const std::size_t __chunk)
+        {
+            // Calculate global diagonal index
+            i_elem = __linear_id * __chunk;
+
+            // Calculate local diagonal index
+            i_elem_local = i_elem % (__n_sorted * 2);
+
+            // Calculate offset to the first element in the subrange (i.e. the first element of the first subrange for merge)
+            offset = std::min<_IndexT>(i_elem - i_elem_local, __n);
+
+            // Calculate size of the first and the second subranges
+            n1 = std::min<_IndexT>(offset + __n_sorted, __n) - offset;
+            n2 = std::min<_IndexT>(offset + __n_sorted + n1, __n) - (offset + n1);
+        }
+
+        inline bool
+        is_i_elem_local_inside_merge_matrix() const
+        {
+            return i_elem_local < n1 + n2;
+        }
+    };
+
+    template <typename Rng>
+    struct DropViews
+    {
+        using __drop_view_simple_t = oneapi::dpl::__ranges::drop_view_simple<Rng, _IndexT>;
+
+        __drop_view_simple_t rng1;
+        __drop_view_simple_t rng2;
+
+        DropViews(Rng& __rng, const WorkDataArea& __data_area)
+            : rng1(__rng, __data_area.offset), rng2(__rng, __data_area.offset + __data_area.n1)
+        {
+        }
+    };
+
+    template <typename _ExecutionPolicy>
+    std::size_t
+    get_max_base_diags_count(_ExecutionPolicy&& __exec, const _IndexT __chunk, std::size_t __n) const
+    {
+        const std::size_t __max_wg_size = oneapi::dpl::__internal::__max_work_group_size(__exec);
+        return oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk * __max_wg_size);
+    }
+
+    // Calculate nd-range params
+    template <typename _ExecutionPolicy>
+    nd_range_params
+    eval_nd_range_params(_ExecutionPolicy&& __exec, const std::size_t __rng_size, const _IndexT __n_sorted) const
+    {
+        const bool __is_cpu = __exec.queue().get_device().is_cpu();
         // The chunk size must not exceed two sorted sub-sequences to be merged,
         // ensuring that at least one work-item processes them.
         const _IndexT __chunk = std::min<_IndexT>(__is_cpu ? 32 : 4, __n_sorted * 2);
-        const std::size_t __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk);
+        const _IndexT __steps = oneapi::dpl::__internal::__dpl_ceiling_div(__rng_size, __chunk);
+
+        _IndexT __base_diag_count = get_max_base_diags_count(__exec, __chunk, __n_sorted);
+        _IndexT __steps_between_two_base_diags = oneapi::dpl::__internal::__dpl_ceiling_div(__steps, __base_diag_count);
+
+        return {__base_diag_count, __steps_between_two_base_diags, __chunk, __steps};
+    }
+
+    template <typename DropViews, typename _Compare>
+    inline static _merge_split_point_t
+    __find_start_point_w(const WorkDataArea& __data_area, const DropViews& __views, _Compare __comp)
+    {
+        return __find_start_point(__views.rng1, decltype(__data_area.n1){0}, __data_area.n1, __views.rng2,
```

Comment:
The type of `__data_area.n1` is known in this context to be `_IndexT`. Can we replace `decltype(__data_area.n1){0}` with `_IndexT{0}`?

Reply:
Fixed.